### ISSUE TITLE
Support truly pluggable protocol/transport for distributed mode

### ISSUE THREAD

**Issue by denismakogon** [tensorflow/tensorflow#15628]
At this moment Distributed TensorFlow supports only one type of transport/protocol which is gRPC, however, it does seem to be configurable (cluster, server, session).

So, there are at least three things that need to be covered:
1. Document the intercommunication protocol (session to a server), i.e. what session sends to a server, what server should respond, etc. The good example here would some kind of swagger spec.
2. Make protocol configurable from Python.
3. Give all a Python interface to implement new types of protocols.

The whole idea is to let developers an ability to implement the protocol that would let distributed TensorFlow spin up and talk a number compute units (serverless, containers, VMs, etc.) instead of having a bunch of processes (cluster servers) running during computations within the session.

Please note, I wasn't able to find any corresponding issues related to given topic.

**Comment by skye** [comment]
FYI @saeta

**Comment by cy89** [comment]
@poxvoculi this seems close to things you work on; can you offer any guidance?
@mrry I think this fits the rubric of distributed training; might we want to expose this kind of functionality? 

Or alternatively, is this relatively open, and should I mark "contributions welcome"?

**Comment by denismakogon** [comment]
@cy89 I do really look for help here. It does seem like changing the whole transport thing is not that simple, so any help is appreciated.

**Comment by poxvoculi** [comment]
@denismakogon I'd like to have a better understanding of what you'd like to be able to do.

Currently TF's runtime execution component is pretty much entirely coded in C++.  Communication between distributed instances generally occurs at this level.  The TF level communication protocol is documented in protocol buffer declarations, e.g.:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/protobuf/worker.proto
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/protobuf/master.proto

Tasks communicate with one another via RPCs that follow these protocols.  The specifics of the RPC system and transport layer don't matter much (other than with respect to performance and reliability :-) ).  gRPC is the only official fully supported transport, but there is support in contrib for sending RecvTensor requests over other RPC interfaces including MPI and verbs.   The intent of the TensorFlow design is to make it relatively easy to substitute another RPC layer by implementing (parts of the) Worker and Master interfaces.  This is configurable at runtime by specifying the any of the available protocols (e.g. 'grpc' but could also be 'grpc+mpi' as [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/mpi/mpi_server_lib.cc#L86)).

That said, what is it that you'd like to do, and what's missing from TF to make that happen?

**Comment by denismakogon** [comment]
@poxvoculi thanks for the reply.

So, here's what i would like to get from distributed mode:

1. Make transport truly pluggable. gRPC is the good thing, but i would like to have HTTP as well. So, i do see the need to give to developers a choice of transports as well as the way to implement transports without recompiling TensorFlow from the source (think of Python interface for transport).

2. Provide task encoder that will accept computation task and return its byte representation to send over the transport. Or make task serializable so the transport layer will decide what to do with that.

So, the feature i would like to have is to let developers decide how would that run their computations within the distributed mode, for instance, i don't want to create the bunch of workers that I would have to maintain by my own processes but delegate it to any orchestration engine (serverless, containers, clear containers, etc.)


Side note
=======
I've found quite interesting note [here](https://www.tensorflow.org/deploy/distributed):
```
Manually specifying these cluster specifications can be tedious, especially for large clusters. We are working on tools for launching tasks programmatically, e.g. using a cluster manager like Kubernetes. If there are particular cluster managers for which you'd like to see support, please raise a GitHub issue.
```

My particular area of the interest is serverless technologies, having Docker or Kubernetes as the backend for the cluster is a good thing, but imagine that get computations units (serverless functions) only when you have something to calculate. So, when i create a computation session (`with tf.Session("http://worker7.example.com:8080") as sess:` where service that stands behind that URL nothing but serverless gateway)  that talks to a serverless function all computations are serialized and send within HTTP request(s) can be deserialized and executed within the serverless function.

**Comment by mrry** [comment]
The `protocol` argument that you provide when creating a `tf.train.Server` and/or in the prefix of the `target` argument to `tf.Session()` provides a way to switch to a different transport protocol. As @poxvoculi notes, we currently support `"grpc"` in the core, but folks have contributed `"grpc+verbs"` and `"grpc+mpi"` implementations of these. It's conceivable that you could use the same extension mechanism for (e.g.) `"http"`. If that's not truly pluggable, please let us know what's missing, and we can then consider the feature request.

I'm not sure what a "task encoder" is in this context. Work items in TensorFlow are typically serialized as `GraphDef` protocol buffers, which I think should be sufficient, but if you could expand on this point, we might understand better.

**Comment by poxvoculi** [comment]
I don't understand what you mean by making transport pluggable.  Communication between TF distributed processes is designed to take place only via methods on the Master and Worker interfaces.  These methods need to be implemented as RPCs.  gRPC is the default RPC implementation but others are possible.  (Inside Google we use an additional two RPC systems, selected among by startup configuration options.  The contrib support cited above are examples of how to do this.)  Unless I'm mistaken, gRPC actually sits on top of HTTP, i.e. the server side uses an HTTP server to field requests.   By 'transport' do you mean a particular implementation of the Master and Worker interfaces, or do you want TF processes to communicate via some completely different methods than are in those interfaces?

**Comment by denismakogon** [comment]
> By 'transport' do you mean a particular implementation of the Master and Worker interfaces, or do you want TF processes to communicate via some completely different methods that are in those > interfaces?

Exactly, I want to make both master and worker use different protocols (anything else different from gRPC). I want to control every send/receive computation request along with being able to serialize/deserialize computation task object (similar to `cloudpickle` but in terms of TF).

gRPC or any other RPC is not the case for me because I don't want to build an additional layer in-between service I run and TF that translates RPC requests into the actual computation task to run.

**Comment by tensorflowbutler** [comment]
A member of the TensorFlow organization has replied after the stat:awaiting tensorflower label was applied.

**Comment by poxvoculi** [comment]
I'm afraid I still don't have a good idea what you're suggesting.   I don't understand what you mean by protocol, since you seem to be rejecting the use of the master.proto and worker.proto interfaces, which are the only way TF processes intercommunicate.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by denismakogon** [comment]
@poxvoculi by saying protocol i meant the actual interaction protocol (who sends what?). At this moment all interactions are being described by protobuf, so there's no way to build another one interaction protocol for all components, for instance, if i want to get rid of protobuf and use, let's say, HTTP 1.1 for that.

>  since you seem to be rejecting the use of the master.proto and worker.proto interfaces, which are the only way TF processes intercommunicate.

This is what i would like to change, i don't like the idea of mandatory protocols, this is very critical to those use cases when people want to build their own workers to achieve massive scale-in/out without exposing too much information about the infrastructure.

**Comment by cy89** [comment]
@denismakogon I agree with @poxvoculi in that I'm still not quite sure what you're proposing. It seems like you can replace gRPC with the RPC package of your choice. I think that if you wanted to replace protobufs with some other portable marshalling solution, you'd have a huge amount of work on your hands for not necessarily a great deal of benefit (I agree that protos can be slow, but faster solutions that retain portability are hard work). 

Thinking constructively, I wonder if a GitHub issue isn't quite the right way to have this discussion. Could we get you to write a somewhat longer-form document that describes (a) the problem you want to address, (b) the limitations of the current solutions, and (c) the way forward that you'd like to explore? It seems to me that a document would allow you to make a reasoned argument for what you want in a way that lets readers better understand where you're coming from.

**Comment by byronyi** [comment]
Just my 2 cents, you can still use HTTP1.1 by serializing protobuf to JSON. I cannot see any benefits doing this though.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by poxvoculi** [comment]
I'm going to close this to stop the auto-nagging.  Feel free to reopen if you want to respond to @cy89 's comment.
